README.md
App-GitHooks
============

[![Build Status](https://travis-ci.org/guillaumeaubert/App-GitHooks.svg?branch=master)](https://travis-ci.org/guillaumeaubert/App-GitHooks)
[![Coverage Status](https://coveralls.io/repos/guillaumeaubert/App-GitHooks/badge.svg?branch=master)](https://coveralls.io/r/guillaumeaubert/App-GitHooks?branch=master)
[![CPAN](https://img.shields.io/cpan/v/App-GitHooks.svg)](https://metacpan.org/release/App-GitHooks)
[![License](https://img.shields.io/badge/license-Perl%205-blue.svg)](http://dev.perl.org/licenses/)

`App::GitHooks` is an extensible and easy to configure git hooks framework that
supports many plugins.


OVERVIEW
--------

 * Here's an example of it in action, running the `pre-commit` hook checks before
the commit message can be entered:

	![Successful checks](https://raw.github.com/guillaumeaubert/App-GitHooks/master/img/app-githooks-example-success.png)

 * Here is another example, with a Perl file that fails compilation this time:

	![Failing checks](https://raw.github.com/guillaumeaubert/App-GitHooks/master/img/app-githooks-example-failure.png)


INSTALLATION
------------

1. Install this distribution (with `cpanm` or your preferred CPAN client):

		cpanm App::GitHooks

2. Install the plugins you are interested in (with `cpanm`or your prefered CPAN
   client), as `App::GitHooks` does not bundle them. See the list of plugins
   below, but for example:

		cpanm App::GitHooks::Plugin::BlockNOCOMMIT
		cpanm App::GitHooks::Plugin::DetectCommitNoVerify
		...

3. Go to the git repository for which you want to set up git hooks, and run:

		githooks install

4. Enjoy!


OFFICIALLY SUPPORTED PLUGINS
----------------------------

 * [App::GitHooks::Plugin::BlockNOCOMMIT]
   (https://metacpan.org/pod/App::GitHooks::Plugin::BlockNOCOMMIT)

Prevent committing code with #NOCOMMIT mentions.

 * [App::GitHooks::Plugin::BlockProductionCommits]
   (https://metacpan.org/pod/App::GitHooks::Plugin::BlockProductionCommits)

Prevent commits in a production environment.

 * [App::GitHooks::Plugin::DetectCommitNoVerify]
   (https://metacpan.org/pod/App::GitHooks::Plugin::DetectCommitNoVerify)

Find out when someone uses --no-verify and append the pre-commit checks to the
commit message.

 * [App::GitHooks::Plugin::ForceRegularUpdate]
   (https://metacpan.org/pod/App::GitHooks::Plugin::ForceRegularUpdate)

Force running a specific tool at regular intervals.

 * [App::GitHooks::Plugin::MatchBranchTicketID]
   (https://metacpan.org/pod/App::GitHooks::Plugin::MatchBranchTicketID)

Detect discrepancies between the ticket ID specified by the branch name and the
one in the commit message.

 * [App::GitHooks::Plugin::PerlCompile]
   (https://metacpan.org/pod/App::GitHooks::Plugin::PerlCompile)

Verify that Perl files compile without errors.

 * [App::GitHooks::Plugin::PerlCritic]
   (https://metacpan.org/pod/App::GitHooks::Plugin::PerlCritic)

Verify that all changes and addition to the Perl files pass PerlCritic checks.

 * [App::GitHooks::Plugin::PerlInterpreter]
   (https://metacpan.org/pod/App::GitHooks::Plugin::PerlInterpreter)

Enforce a specific Perl interpreter on the first line of Perl files.

 * [App::GitHooks::Plugin::PgBouncerAuthSyntax]
   (https://metacpan.org/pod/App::GitHooks::Plugin::PgBouncerAuthSyntax)

Verify that the syntax of PgBouncer auth files is correct.

 * [App::GitHooks::Plugin::PrependTicketID]
   (https://metacpan.org/pod/App::GitHooks::Plugin::PrependTicketID)

Derive a ticket ID from the branch name and prepend it to the commit-message.

 * [App::GitHooks::Plugin::RequireCommitMessage]
   (https://metacpan.org/pod/App::GitHooks::Plugin::RequireCommitMessage)

Require a commit message.

 * [App::GitHooks::Plugin::RequireTicketID]
   (https://metacpan.org/pod/App::GitHooks::Plugin::RequireTicketID)

Verify that staged Ruby files compile.

 * [App::GitHooks::Plugin::ValidatePODFormat]
   (https://metacpan.org/pod/App::GitHooks::Plugin::ValidatePODFormat)

Validate POD format in Perl and POD files.


CONTRIBUTED PLUGINS
-------------------

 * [App::GitHooks::Plugin::RubyCompile]
   (https://metacpan.org/pod/App::GitHooks::Plugin::RubyCompile)

Verify that staged Ruby files compile.

 * [App::GitHooks::Plugin::PreventTrailingWhitespace]
   (https://metacpan.org/pod/App::GitHooks::Plugin::PreventTrailingWhitespace)

Prevent trailing whitespace from being committed.


SUPPORT AND DOCUMENTATION
-------------------------

After installing, you can find documentation for this module with the
perldoc command.

	perldoc App::GitHooks


You can also look for information at:

 * [GitHub's request tracker (report bugs here)]
   (https://github.com/guillaumeaubert/App-GitHooks/issues)

 * [AnnoCPAN, Annotated CPAN documentation]
   (http://annocpan.org/dist/App-GitHooks)

 * [CPAN Ratings]
   (http://cpanratings.perl.org/d/App-GitHooks)

 * [MetaCPAN]
   (https://metacpan.org/release/App-GitHooks)


LICENSE AND COPYRIGHT
---------------------

Copyright (C) 2013-2016 Guillaume Aubert.

This code is free software; you can redistribute it and/or modify it under the
same terms as Perl 5 itself.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the LICENSE file for more details.


Changes
Revision history for App-GitHooks

v1.8.0  2016-05-21
        - Switched to Test::Requires::Git instead of the now obsolete
          Test::Git::has_git(). Thank you, Philippe Bruhat!
        - Added App::GitHooks->get_remote_name() to get the name of the
          repository.
        - Fixed typo in documentation. Thank you, Yanick Champoux!
        - Relicensed under Perl 5 terms.

v1.7.3  2015-06-22
        - Removed relative path include in ./bin/githooks (GH-25).
        - Minor code and documentation cleanup.

v1.7.2  2015-04-08
        - Retrieve stdin at the hook level for pre-push.
        - Catch plugin exceptions.

v1.7.1  2015-04-07
        - Added pre-push hook file to the hooks directory (GH-22).

v1.7.0  2015-04-06
        - Implemented support for pre-push hook (GH-22).

v1.6.1  2015-03-31
        - Fixed test that prevented upgrading App::GitHooks in situations where
          plugins were already installed and returned a different value than
          expected in the test.

v1.6.0  2015-03-30
        - Added App::GitHooks::Test::ok_reset_githooksrc() to ensure that tests
          can run without interference.
        - Refactored tests to use ok_reset_githooksrc().
        - Added .githooksrc to manage App::GitHooks' commits with App::GitHooks
          (GH-21).
        - Implement support for warnings in the pre-commit hook (GH-13).

v1.5.1  2015-03-29
        - Account for git-commit trimming whitespace in commit messages
          (GH-20).

v1.5.0  2015-03-29
        - Rename App::GitHooks::Utils::get_ticket_id() to
          App::GitHooks::Utils::get_ticket_id_from_commit_regex() and fix the
          function to match its documentation (GH-19).

v1.4.0  2015-03-27
        - Implement "force_plugins" configuration option (GH-18).

v1.3.0  2015-03-25
        - Add "list" command to "githooks" to display available plugins for
          each git hook (GH-8).

v1.2.1  2015-03-23
        - Replace File::Slurp with Path::Tiny (GH-17).
        - Improve and clean up tests.
        - Improve and clean up documentation.
        - Make constants read-only.

v1.2.0  2015-01-21
        - Fix running git actions with an underscore in their name and that are
          using the default App::GitHooks::Hook->run() (GH-15). Thank you,
          Jacob Maurer!

v1.1.5  2014-09-06
        - Fix documentation typos.

v1.1.4  2014-08-22
        - Added missing "autodie" pre-requisite.

v1.1.3  2014-08-16
        - Documented the new "githooks" command and simplified the synopsis.

v1.1.2  2014-08-13
        - Require Test::Compile v1.1.0 to use its object-oriented interface.

v1.1.1  2014-08-11
        - Amended tests to use the current Perl to instantiate "githooks".

v1.1.0  2014-08-10
        - Added "githooks" command line utility to set up in git repositories
          all the hooks supported by App::GitHooks.

v1.0.8  2014-06-11
        - Forced encoding only when utf8 characters are to be printed (part 2).

v1.0.7  2014-06-07
        - Forced encoding only when utf8 characters are to be printed.
        - Added screenshots of App::GitHooks in action.
        - Improved information in readme file.
        - Fixed interpreter line in examples and hooks/.
        - Noted the git version in tests to help troubleshooting.

v1.0.6  2014-06-03
        - Bumped git requirement to v1.7.4.1.
        - Documented git requirement.

v1.0.5  2014-06-02
        - Synced manifest.

v1.0.4  2014-06-02
        - Added git v1.5.0 as a requirement for the tests.

v1.0.3  2014-06-01
        - Fixed test coverage when using 'cover' instead of 'prove'.

v1.0.2  2014-05-31
        - Fixed prerequisites.

v1.0.1  2014-05-31
        - Tweaked POD formatting.
        - Instantiated a new Git::Repository object on demand only to help with
          unit testing.

v1.0.0  2014-05-30
        - Public release.


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.4205
require 5.010;
use ExtUtils::MakeMaker;
WriteMakefile
(
  'PL_FILES' => {},
  'VERSION_FROM' => 'lib/App/GitHooks.pm',
  'EXE_FILES' => [
                   'bin/githooks'
                 ],
  'test' => {
              'TESTS' => 't/*.t t/45-Hook/*.t t/45-Hook/PreCommit/*.t t/45-Hook/PrepareCommitMsg/*.t t/45-Hook/CommitMsg/*.t t/17-CommitMessage/*.t t/14-Utils/*.t t/13-Config/*.t t/90-githooks/*.t t/05-Terminal/*.t t/40-Plugin/*.t t/40-Plugin/PrintSTDERR/*.t t/40-Plugin/CustomReply/*.t t/06-Constants/*.t'
            },
  'PREREQ_PM' => {
                   'Cwd' => 0,
                   'Test::Git' => 0,
                   'Term::ANSIColor' => 0,
                   'Try::Tiny' => 0,
                   'Data::Validate::Type' => 0,
                   'Class::Load' => 0,
                   'Scalar::Util' => 0,
                   'Path::Tiny' => 0,
                   'Term::Encoding' => 0,
                   'Test::Compile' => '1.001',
                   'Test::FailWarnings' => 0,
                   'Readonly' => 0,
                   'File::Spec' => 0,
                   'autodie' => 0,
                   'Module::Pluggable' => 0,
                   'Capture::Tiny' => 0,
                   'Git::Repository' => 0,
                   'Test::More' => 0,
                   'Test::Exception' => 0,
                   'Data::Section' => 0,
                   'Data::Dumper' => 0,
                   'Config::Tiny' => 0,
                   'Storable' => 0,
                   'Test::Type' => '1.002',
                   'Getopt::Long' => 0,
                   'Carp' => 0,
                   'File::Basename' => 0,
                   'Test::Requires::Git' => '1.005',
                   'Parallel::ForkManager' => 0,
                   'File::Temp' => 0,
                   'Pod::Usage' => 0,
                   'Term::ReadKey' => 0,
                   'Text::Wrap' => 0,
                   'Test::Deep' => 0
                 },
  'NAME' => 'App::GitHooks',
  'INSTALLDIRS' => 'site'
)
;
